<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Problem-1" data-toc-modified-id="Problem-1-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Problem 1</a></span><ul class="toc-item"><li><span><a href="#a)-Formulate-a-linear-programming-model-for-this-problem-to-maximize-profit" data-toc-modified-id="a)-Formulate-a-linear-programming-model-for-this-problem-to-maximize-profit-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>a) Formulate a linear programming model for this problem to maximize profit</a></span></li><li><span><a href="#b)-Use-the-graphical-method-to-solve-this-model" data-toc-modified-id="b)-Use-the-graphical-method-to-solve-this-model-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>b) Use the graphical method to solve this model</a></span></li><li><span><a href="#c)-Solve-with-Python" data-toc-modified-id="c)-Solve-with-Python-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>c) Solve with Python</a></span></li><li><span><a href="#d)-What-constraints-are-binding?" data-toc-modified-id="d)-What-constraints-are-binding?-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>d) What constraints are binding?</a></span></li></ul></li><li><span><a href="#Problem-2" data-toc-modified-id="Problem-2-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Problem 2</a></span><ul class="toc-item"><li><span><a href="#a)-Use-graphical-analysis-to-determine-the-optimal-solution(s)" data-toc-modified-id="a)-Use-graphical-analysis-to-determine-the-optimal-solution(s)-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>a) Use graphical analysis to determine the optimal solution(s)</a></span></li><li><span><a href="#b)-Simplex-Algorithm" data-toc-modified-id="b)-Simplex-Algorithm-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>b) Simplex Algorithm</a></span></li></ul></li><li><span><a href="#Problem-3" data-toc-modified-id="Problem-3-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Problem 3</a></span><ul class="toc-item"><li><span><a href="#a)--Report-the-optimal-solution-for-decision-variables-and-the-objective-value." data-toc-modified-id="a)--Report-the-optimal-solution-for-decision-variables-and-the-objective-value.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>a)  Report the optimal solution for decision variables and the objective value.</a></span></li><li><span><a href="#b)-Report-and-interpret-the-reduced-cost-value-of-each-decision-variable." data-toc-modified-id="b)-Report-and-interpret-the-reduced-cost-value-of-each-decision-variable.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>b) Report and interpret the reduced cost value of each decision variable.</a></span></li></ul></li><li><span><a href="#Problem-4" data-toc-modified-id="Problem-4-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Problem 4</a></span><ul class="toc-item"><li><span><a href="#a)-Formulate-an-LP-model-and-implement-it-in-Python." data-toc-modified-id="a)-Formulate-an-LP-model-and-implement-it-in-Python.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>a) Formulate an LP model and implement it in Python.</a></span></li><li><span><a href="#b)-What-is-the-optimal-solution?" data-toc-modified-id="b)-What-is-the-optimal-solution?-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>b) What is the optimal solution?</a></span></li><li><span><a href="#c)-What-constraints-are-binding?-Interpret-the-shadow-prices-and-slack-values." data-toc-modified-id="c)-What-constraints-are-binding?-Interpret-the-shadow-prices-and-slack-values.-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>c) What constraints are binding? Interpret the shadow prices and slack values.</a></span></li><li><span><a href="#d)-Interpret-the-reduced-cost-value-of-each-decision-variable." data-toc-modified-id="d)-Interpret-the-reduced-cost-value-of-each-decision-variable.-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>d) Interpret the reduced cost value of each decision variable.</a></span></li></ul></li><li><span><a href="#Problem-5" data-toc-modified-id="Problem-5-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Problem 5</a></span><ul class="toc-item"><li><span><a href="#a)-Formulate-an-Integer-LP-model-for-this-problem." data-toc-modified-id="a)-Formulate-an-Integer-LP-model-for-this-problem.-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>a) Formulate an Integer LP model for this problem.</a></span></li><li><span><a href="#b)-Implement-the-model-in-Python-and-solve-it." data-toc-modified-id="b)-Implement-the-model-in-Python-and-solve-it.-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>b) Implement the model in Python and solve it.</a></span></li></ul></li><li><span><a href="#Problem-6" data-toc-modified-id="Problem-6-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Problem 6</a></span><ul class="toc-item"><li><span><a href="#a)-Formulate-a-BIP-(Binary-Integer-Programming)-model-for-this-problem." data-toc-modified-id="a)-Formulate-a-BIP-(Binary-Integer-Programming)-model-for-this-problem.-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>a) Formulate a BIP (Binary Integer Programming) model for this problem.</a></span></li><li><span><a href="#b)-Solve-in-Python" data-toc-modified-id="b)-Solve-in-Python-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>b) Solve in Python</a></span></li></ul></li><li><span><a href="#Problem-7" data-toc-modified-id="Problem-7-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Problem 7</a></span><ul class="toc-item"><li><span><a href="#a)-Formulate-a-MIP-(Mixed-Integer-Programming)-model" data-toc-modified-id="a)-Formulate-a-MIP-(Mixed-Integer-Programming)-model-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>a) Formulate a MIP (Mixed Integer Programming) model</a></span></li><li><span><a href="#b)-Solve-in-Python" data-toc-modified-id="b)-Solve-in-Python-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>b) Solve in Python</a></span></li></ul></li></ul></div>

## Problem 1 

The WorldLight Company produces two light fixtures (products 1 and 2) that require both metal frame parts and electrical components. Management wants to determine how many units of each product to produce so as to maximize profit.

- For each unit of product 1, 1 unit of frame parts and 2 units of electrical components are required.
- For each unit of product 2, 3 units of frame parts and 2 units of electrical components are required.
- The company has 200 units of frame parts and 300 units of electrical components.
- Each unit of product 1 gives a profit of \\$1, and each unit of product 2, up to 60 units, gives a profit of \\$2.
- Any excess over 60 units of product 2 brings no profit, so such an excess has been ruled out.

### a) Formulate a linear programming model for this problem to maximize profit

$Z = x_1+2x_2$

$x_1 + 3x_2 \leq 200$

$2x_1 + 2x_2 \leq 300$

$x_2 \leq 60$

### b) Use the graphical method to solve this model
<img src="Images/HW 1-1b.png">

### c) Solve with Python
>Compare your answer with part b. Also, report the sensitivity analysis parameters such as slack and shadow prices for all constraints.

Below we create our model, define the objective function, input the constraints, and solve. The results conflict with the solution obtained above graphically, and it appears this is due to choosing the incorrect cornerpoint. 

In [1]:
from pulp import *

In [9]:
# Model with lower bounds of zero
prob = LpProblem("WorldLight Co", LpMaximize)
x1 = LpVariable('x_1', lowBound = 0)
x2 = LpVariable('x_2', lowBound = 0)

# Profit/objective function
prob += x1 + 2*x2, "Obj"

# Constraints
prob += x2 <= 60 # Product 2 upper bound
prob += x1 + 3*x2 <= 200 # frame parts
prob += 2*x1 + 2*x2 <= 300 # components

# Solve the model
prob.solve()

# Loop to print the optimum variables 
for variable in prob.variables():
    print("{}* = {}".format(variable.name, variable.varValue))

# Profit at optimum levels
print(value(prob.objective))

x_1* = 125.0
x_2* = 25.0
175.0


The result finds the optimum values of 125 of product 1 and 25 of product 2 for a profit amount of $175.

To continue our analysis, we can find the slack values and shadow prices for each constraint:

In [11]:
for name, c in prob.constraints.items():
    print("\n", name, ":", c, ", Slack=", c.slack, ", Shadow Price=", c.pi)


 _C1 : x_2 <= 60 , Slack= 35.0 , Shadow Price= -0.0

 _C2 : x_1 + 3*x_2 <= 200 , Slack= -0.0 , Shadow Price= 0.5

 _C3 : 2*x_1 + 2*x_2 <= 300 , Slack= -0.0 , Shadow Price= 0.25


### d) What constraints are binding?
Interpret the shadow price value of each constraint.

The constraints for the frame parts and components are binding, and the shadow prices reflect how much more profit could be obtained if these constraints are increased by one. The constraint of product 2 has a shadow price of zero, meaning this is a free good and is not binding.

## Problem 2
Consider the following problem, where the value of 𝑐 1 has not yet been ascertained.

$Z = c_1x_1 +x_2$

$x_1+x_2 \leq 6$

$x_1+2x_2 \leq 10$

$x_1,x_2 \geq 0$

### a) Use graphical analysis to determine the optimal solution(s)
...for ( 𝑥1 , 𝑥2 ) for the various possible values of $𝑐_1$ . Try 3 different values (for instance it can be 0.5, 1, and 2). Have you seen any change on the optimal solution of $𝑥_1 , 𝑥_2$ and objective value? Explain.

### b) Simplex Algorithm
Suppose $c_1 = 2$ (so $Z=2x_1+x_2$, solve with the simplex model.

## Problem 3

Here we explore a simple LP minimization model with the following constraints:

$Z = 4x_1 + 3_x2 + 5x_3$

$2x_1 + x_3 \geq 8$

$x_2 + x_3 \geq 6$

$6x_1 + 8x_2 \geq 48$

$x_1,x_2,x_3 \geq 0$

### a)  Report the optimal solution for decision variables and the objective value.

Performing a minimization is extremely similar to the maximization process in Problem 1. It is a simple matter of replacing the parameter when calling the LP Problem class. The optimal results for $x_1,x_2,x_3$ are printed below, which result in a minimized objective value of 34.

In [14]:
# Create problem class and variables
from pulp import *
prob = LpProblem("LP Minimization", LpMinimize)
x1 = LpVariable('x_1', lowBound = 0)
x2 = LpVariable('x_2', lowBound = 0)
x3 = LpVariable('x_3', lowBound = 0)

# Objective function 
prob += 4*x1 + 3*x2 + 5*x3, "Obj"

# Constraints
prob += 2*x1 + x3 >= 8
prob += x2 + x3 >= 6
prob += 6*x1 + 8*x2 >= 48
prob += x1 >= 0
prob += x2 >= 0
prob += x3 >= 0

# Solve the minimization
prob.solve()

print(LpStatus[prob.status])

for variable in prob.variables():
    print("{} = {}".format(variable.name, variable.varValue))
    
print("Z* = ", value(prob.objective))

Optimal
x_1 = 2.9090909
x_2 = 3.8181818
x_3 = 2.1818182
Z* =  34.0


### b) Report and interpret the reduced cost value of each decision variable.

The reduced cost can be obtained from the resulting prob.variables object. In this case the reduced cost of each decision variable is zero, which makes sense given that all three optimized variables have positive values. If there were a reduced cost, it would give us the amount that the coefficient of the variable in our objective function would need to increase before it would have a positive optimal value.  

In [15]:
for v in prob.variables():
    print ("\n", v.name, "=", v.varValue, ", Reduced Cost=", v.dj)


 x_1 = 2.9090909 , Reduced Cost= 0.0

 x_2 = 3.8181818 , Reduced Cost= 0.0

 x_3 = 2.1818182 , Reduced Cost= 0.0


## Problem 4
> A national bank needs to determine how to invest \\$500,000 in the following collection of bonds to maximize the annual return:


| Bond | Annual Return | Maturity | Risk | Tax-Free |
|------|---------------|----------|------|----------|
| A    | 10.0%         | Long     | High | Yes      |
| B    | 7.5%          | Short    | Low  | Yes      |
| C    | 9.0%          | Long     | Low  | No       |
| D    | 9.0%          | Long     | High | Yes      |
| E    | 9.0%          | Short    | High | No       |

>The officer wants to invest at least 50\% of the money in short-term issues and no more than 50% in high-risk issues. Also, at least 30% of the funds should go into tax-free investments.

### a) Formulate an LP model and implement it in Python.

To create this model, we start by creating the problem class and carefully choosing upward bounds that correspond to the prompt. In this case, bonds A, D, and E are high-risk and can only receive 50% of the investment so we set `upBound = 250`. We are maximizing the annual return, so we use those return percentages to build out our model. 

Constraints include:
- B and D should sum to $\geq$ than 250 as 50% short-term bonds
- A, C, and D should sum to $\leq$  than 250 as 50% high-risk bonds
- A, B, and D should sum to $\geq$  150 as 30% tax-free bonds

In [19]:
prob = LpProblem("My LP Problem", LpMaximize)

x1 = LpVariable('A', lowBound = 0, upBound = 250)
x2 = LpVariable('B', lowBound = 0, upBound = 500)
x3 = LpVariable('C', lowBound = 0, upBound = 500)
x4 = LpVariable('D', lowBound = 0, upBound = 250)
x5 = LpVariable('E', lowBound = 0, upBound = 250)

# Objective function based on annual return
prob += 0.10*x1 + 0.075*x2 + 0.09*x3 + 0.09*x4 + 0.09*x5, "Obj"

# Constraints
prob += x1 + x2 + x3 + x4 + x5 == 500, "Portfolio" 
prob += x2 + x5 >= 250, "Short-Term"
prob += x1 + x3 + x4 <= 250, "High-Risk"
prob += x1 + x2 + x4 >= 150, "Tax-Free"

print(prob)

prob.solve() 
print("status: " + LpStatus[prob.status]) 

for variable in prob.variables():
     print("{}* = {}".format(variable.name, variable.varValue)) 
     
print("Z* = ",value(prob.objective))

My_LP_Problem:
MAXIMIZE
0.1*A + 0.075*B + 0.09*C + 0.09*D + 0.09*E + 0.0
SUBJECT TO
Portfolio: A + B + C + D + E = 500

Short_Term: B + E >= 250

High_Risk: A + C + D <= 250

Tax_Free: A + B + D >= 150

VARIABLES
A <= 250 Continuous
B <= 500 Continuous
C <= 500 Continuous
D <= 250 Continuous
E <= 250 Continuous

status: Optimal
A* = 250.0
B* = 0.0
C* = 0.0
D* = 0.0
E* = 250.0
Z* =  47.5


### b) What is the optimal solution?

The optimal solution is to split the investment evenly between bonds A and E. This results in an optimal annual return of $47,500.

### c) What constraints are binding? Interpret the shadow prices and slack values.

The portfolio, short-term, and high-risk constraints are binding. The shadow price of the portfolio constraint is unsurprising, increasing the amount of investment would increase the objective value. The shadow price of the short-term constraint implies a reduced return if we increased its investment. The tax-free constraint is non-binding and has a slack of -100, meaning we have exceeded our tax-free requirement by 100. 

In [20]:
for name, c in prob.constraints.items():
    print("\n{}: Slack={}, Shadow Price = {}".format(name, c.slack, c.pi))
    print("\n", name, ":", ", Slack=", c.slack, ", Shadow Price=", c.pi)


Portfolio: Slack=-0.0, Shadow Price = 0.1

 Portfolio : , Slack= -0.0 , Shadow Price= 0.1

Short_Term: Slack=-0.0, Shadow Price = -0.025

 Short_Term : , Slack= -0.0 , Shadow Price= -0.025

High_Risk: Slack=-0.0, Shadow Price = -0.0

 High_Risk : , Slack= -0.0 , Shadow Price= -0.0

Tax_Free: Slack=-100.0, Shadow Price = -0.0

 Tax_Free : , Slack= -100.0 , Shadow Price= -0.0


### d) Interpret the reduced cost value of each decision variable.

The reduced cost of A and B are zero, which implies that B is not present in the optimum solution due to a combination of constraints that ruled it out. The negative reduced costs are strange for a maximization problem, it might imply these are actually better investments than the optimum solution but are ruled out due to constraints. The positive reduced cost of E, a bond that is present in the solution, may strengthen that argument because it would mean that it should be increased by 0.015 to be optimal.

In [21]:
for variable in prob.variables():
     print("\n{}* = {},   Reduced Cost = {}".format(variable.name, variable.varValue, variable.dj)) 


A* = 250.0,   Reduced Cost = 0.0

B* = 0.0,   Reduced Cost = 0.0

C* = 0.0,   Reduced Cost = -0.01

D* = 0.0,   Reduced Cost = -0.01

E* = 250.0,   Reduced Cost = 0.015


## Problem 5
>The CitruSun Corporation ships frozen orange juice concentrate from processing plants in Eustis and Clermont to distributors in Miami, Orlando, and Tallahassee. Each plant can produce 20 tons of concentrate each week. The company has just received orders of 10 tons from Miami for the coming week, 15 tons for Orlando, and 10tons for Tallahassee. The cost per ton for supplying each of the distributors from each of the processing plants i s shown in the following table:

|          | Miami | Orlando | Tallahassee |
|----------|-------|---------|-------------|
| Eustis   | \$260  | \$200    | \$290        |
| Clermont | \$220  | \$240    | $330        |

>The company wants to determine the **minimum** costly plan for filling their orders for the coming week.

### a) Formulate an Integer LP model for this problem.

### b) Implement the model in Python and solve it.
> What is the optimal solution?

## Problem 6

A couple wants to divide their household tasks between them so that each has two tasks but the time spent is kept to a minimum. Their efficiencies on this task differ, and the time needed for each task/person is given in this table in hours.

|        | Shopping | Cooking | Dishwashing | Laundry |
|--------|----------|---------|-------------|---------|
| Eve    | 4.5      | 7.5     | 3.0         | 3.0     |
| Steven | 5.5      | 7.2     | 4.5         | 3.2     |

### a) Formulate a BIP (Binary Integer Programming) model for this problem.
(Needs 8 binary decision variables)

### b) Solve in Python

## Problem 7

The Toys-R-4-U Company has developed two new toys for possible inclusion in its product line for the upcoming Christmas season. Setting up the production facilities to begin production would cost $45,000 for toy 1 and $75,000 for toy 2. Once these costs are covered, the toys would generate a unit profit of $10 for toy 1 and $15 for toy 2.

The company has two factories that are capable of producing these toys. However, to avoid doubling the start-up costs, just one factory would be used, where the choice would be based on maximizing profit. For administrative reasons, the same factory would be used for both new toys if both are produced.

Toy 1 can be produced at the rate of 50 per hour (or 1/50 hours for each toy) in factory 1 and 40 per hour (1/40 hours for each toy) in factory 2. Toy 2 can be produced at the rate of 40 per hour (1/40 hours for each toy) in factory 1 and 25 per hour in factory 2. Factories 1 and 2, respectively, have 500 hours and 700 hours of production time available before Christmas that could be used to produce these toys.

It is not known whether these two toys would be continued after Christmas. Therefore, the problem is to determine how many units (if any) of each new toy should be produced before Christmas to maximize the total profit.

### a) Formulate a MIP (Mixed Integer Programming) model

### b) Solve in Python